src/mnode/mnode-pastel.h

```diff
@@ -618,3 +619,1814 @@ class CTakeDownTicket : public CPastelTicket
 
     void SerializationOp(CDataStream& s, const SERIALIZE_ACTION ser_action) override {}
 };
+
+// Username Change Ticket /////////////////////////////////////////////////////////////////////////////////////////////////////
+/*
+	"ticket": {
+		"type": "username",
+		"pastelID": "",    //PastelID of the username
+		"username": "",    //new valid username
+		"fee": "",         // fee to change username
+		"signature": ""
+	},
+ */
+class CChangeUsernameTicket : public CPastelTicket
+{
+public:
+    std::string pastelID;
+    std::string username;
+    CAmount fee{100};
+    std::vector<unsigned char> signature;
+
+public:
+    CChangeUsernameTicket() = default;
+
+	explicit CChangeUsernameTicket(std::string _pastelID) :
+        pastelID(std::move(_pastelID))
+    {}
+    std::string KeyOne() const noexcept override { return username; }
+    std::string KeyTwo() const noexcept override { return pastelID; }
+
+    bool HasKeyTwo() const noexcept override { return true; }
+    bool HasMVKeyOne() const noexcept override { return false; }
+    bool HasMVKeyTwo() const noexcept override { return false; }
+
+    void SetKeyOne(std::string val) override { username = std::move(val); }
+
+    static CChangeUsernameTicket Create(std::string _pastelID, std::string _username, const SecureString& strKeyPass);
+    
+    static bool FindTicketInDb(const std::string& key, CChangeUsernameTicket& ticket);
+    
+    CAmount TicketPrice(const unsigned int nHeight) const noexcept override { return fee; }
+
+    TicketID ID() const noexcept override { return TicketID::Username; }
+    static TicketID GetID() { return TicketID::Username; }
+
+    std::string ToJSON() const noexcept override;
+    std::string ToStr() const noexcept override;
+
+    bool IsValid(bool preReg, int depth) const override;
+
+    void SerializationOp(CDataStream& s, const SERIALIZE_ACTION ser_action) override
+    {
+        const bool bRead = ser_action == SERIALIZE_ACTION::Read;
+        std::string error;
+        if (!VersionMgmt(error, bRead))
+            throw std::runtime_error(error);
+        READWRITE(pastelID);
+        READWRITE(m_nVersion);
+        // v0
+        READWRITE(username);
+        READWRITE(fee);
+        READWRITE(signature);
+        READWRITE(m_nTimestamp);
+        READWRITE(m_txid);
+        READWRITE(m_nBlock);
+    }
+
+    /** Some general checks to see if the username is bad. Below cases will be considered as bad Username
+    *     - Contains characters that is different than upper and lowercase Latin characters and numbers
+    *     - Has only <4, or has more than 12 characters
+    *     - Doesn't start with letters.
+    *     - Username registered on the blockchain.
+    *     - Contains bad words (swear, racist,...)
+    * return: true if bad, false if good to use
+    */
+    static bool isUsernameBad(const std::string& username, std::string& error);
+
+};
+
+struct UsernameBadWords {
```

Comment:
This is created to serve UsernameChange ticket only, so I think it is better to place it here.
Using Singleton is better than just a function with staic list initialized because of **testability**. We can do unit test easily with Singleton, but we can't achieve it easily with a function with static list initialized at first call.

Reply:
mnode_pastel.h is very big in length due to this function, need to scroll a lot. Using like functions usualy move to some utils.
you can place the singleton in the unit tests if you need to use it. do you use it now in tests?

Reply:
Oh so we dont have to scroll util file if we move it to util?
And singleton is better for testabilities but now we didnt write tests for it yet, so we should use a worst version (function with static list). So in the future if we need ot test it we change back to Singleton?

Reply:
it would be beneficial also later on to split mnode-pastel.h|cpp into separate cpp|h files for each ticket type.
Much easier to merge

Reply:
will fix

Reply:
fixed